### PR TITLE
Add logging to reveal reason to persist the hydrants

### LINKED ISSUE #16408: Stream ingestion persist reason to be revealed
### Description

It would be very useful to know the reason why persist is schedule. There are three possibilities
- maxRowInMemory
- maxBytesInMemory
- nextFlush time
Understanding the reason would be necessary to tune the parameters accordingly 

### Motivation

As stated above, need a signal to tune the ingestion spec.

-----
The proposed implementation would be very straight forward. Aka, add a line of logging.

### PR DESCRIPTION
**Pull request by kaisun2000**
Fixes #16408 16408

### Description

Add a line of logging to give a signal as why persist is going to be triggered next in the peon. This would be useful to tune the ingestion spec. 

The possible reasons are:
- maxRowInMemory reached
- maxBytesInMemory reached
- nextFlush time reached

Currently only "IncreamentalPersist" code path has this logging. Here we added the logging for the rest code path when increamental persist is not enabled.

#### Release note

<hr>

##### Key changed/added classes in this PR
 * `MyFoo`
 * `OurBar`
 * `TheirBaz`

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderator.java [hunk 1]
```diff
@@ -465,6 +465,7 @@ public void onFailure(Throwable t)
             MoreExecutors.directExecutor()
         );
       } else {
+        log.info("Set persist flag for the reason: %s.", String.join(",", persistReasons));
```
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderator.java:468]: Line 402 already has a similar log line. You just need to move it out of the `if`, i.e. move it to line 400.

Comment (kaisun2000) [reply server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderator.java:468]: @kfaraz ping

Comment (kfaraz) [reply server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderator.java:468]: I think it makes sense to have the distinction. But the log line should reflect that this is the non-incremental code path.

```suggestion
        log.info("Marking ready for non-incremental async persist due to reasons[%s].", persistReasons);
```

**Comment by Pankaj260100** [comment]
>The possible reasons are:
maxRowInMemory reached
maxBytesInMemory reached
nextFlush time reached

It also persist all the data to disk when the ingestion worker wants to publish the data to deep storage.

File: server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderator.java (hunk 1 above)
Comment (kaisun2000) [reply server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderator.java]: Updated the logging message to reflect this is non-incremental async path. 
@kfaraz

**Review by kfaraz — APPROVED** [review]

**Pull request merged**